redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt

```diff
@@ -26,101 +26,197 @@ import app.cash.redwood.layout.api.Constraint
+
+  private val viewPortListCoordinator = object {
+    var minIndex: Int = 0
+    var maxIndex: Int = 0
+
+    fun notifyViewportChanged() {
+      onViewportChanged(minIndex, maxIndex)
+    }
+
+    fun updateViewport(minIndex: Int, maxIndex: Int) {
+      if (minIndex != this.minIndex || maxIndex != this.maxIndex) {
+        this.minIndex = minIndex
+        this.maxIndex = maxIndex
+
+        notifyViewportChanged()
       }
     }
+  }
+
+  // A callback to tell LazyList that we have an update to the window of items available
+  private lateinit var onViewportChanged: (firstVisibleItemIndex: Int, lastVisibleItemIndex: Int) -> Unit
 
-  private val tableViewDataSource: UITableViewDataSourceProtocol =
-    object : NSObject(), UITableViewDataSourceProtocol {
-      override fun numberOfSectionsInTableView(
-        tableView: UITableView,
-      ): NSInteger = 1L
-
-      override fun tableView(
-        tableView: UITableView,
-        numberOfRowsInSection: NSInteger,
-      ): NSInteger = itemsList.size.toLong()
-
-      override fun tableView(
-        tableView: UITableView,
-        cellForRowAtIndexPath: NSIndexPath,
-      ): UITableViewCell = tableView.dequeueReusableCellWithIdentifier(reuseIdentifier) as Cell
+  // UICollectionView + Protocols
+  private var collectionViewFlowLayout: UICollectionViewFlowLayout =
+    object: UICollectionViewFlowLayout() {}
+
+  private val collectionViewDataSource: UICollectionViewDataSourceProtocol =
+    object : NSObject(), UICollectionViewDataSourceProtocol {
+
+      override fun collectionView(
+        collectionView: UICollectionView,
+        numberOfItemsInSection: NSInteger,
+      ): NSInteger {
+        return items.itemCount().toLong()
+      }
+
+      override fun collectionView(
+        collectionView: UICollectionView,
+        cellForItemAtIndexPath: NSIndexPath,
+      ): UICollectionViewCell {
+        val cell = collectionView.dequeueReusableCellWithReuseIdentifier(
+          identifier = reuseIdentifier,
+          forIndexPath = cellForItemAtIndexPath
+        ) as LazyListContainerCell
+
+        val view = items.itemForGlobalIndex(cellForItemAtIndexPath.item.toInt())
+        cell.set(view)
+
+        return cell
+      }
     }
 
-  internal val tableView = UITableView()
-    .apply {
-      dataSource = tableViewDataSource
-      delegate = tableViewDelegate
-      rowHeight = 64.0 // TODO: size rows by their content.
-      prefetchingEnabled = false
+  private val collectionViewDelegate: UICollectionViewDelegateFlowLayoutProtocol =
+    object: NSObject(), UICollectionViewDelegateFlowLayoutProtocol {
+      override fun collectionView(
+        collectionView: UICollectionView,
+        layout: UICollectionViewLayout,
+        sizeForItemAtIndexPath: NSIndexPath
+      ): CValue<CGSize> {
+        // TODO: Calculate size for individual items. This commented code currently hangs the UI Thread
```

Comment:
What do you think will be the workaround here? Since we're implementing a synchronous method on the protocol, we can't really be hopping around to different threads.
Would calling `sizeToFit` (instead of `sizeThatFits`) here, or within `cellForItemAtIndexPath` work, or is the issue that this is always going to be called on a background thread?